Add build support for Ubuntu 24.04

nickbianco opened the pull request:
Fixes issue #4184 

### Brief summary of changes

Some of these changes are drive-by improvements, mostly to the CI system.

- Upgraded Python version to 3.11 (with NumPy 2.0) on CI.
- Removed "perf" tests on CI. These were a good idea at the time, but rarely used. [I've saved a copy of the existing CI workflow in the opensim-perf repo for future reference](https://github.com/opensim-org/opensim-perf/blob/main/external_workflows/continuous_integration.yml).
- Added Ubuntu 24.04 CI job. 
- Updated Windows CI job to use more recent `windows-2025` runner.
- I've updated the Simbody commit and `Mat.h` and `Vec.h` under `Bindings/SWIGSimTK` based on @adamkewley's [recent Simbody changes](https://github.com/simbody/simbody/pull/840). This change is necessary to avoid a warning that newer compilers (e.g., GCC 13) about potential out-of-bounds errors in `SimTK::Mat` constructors.
- Converted references to copies in `testOutputReporter.cpp` to avoid dangling reference errors from GCC.
- Changed how data members in `WrapResult` and `PathWrapPoint` are initialized to avoid "possibly uninitialized" errors from GCC.

### Testing I've completed

Ran tests locally and on CI.

### Looking for feedback on...

### CHANGELOG.md (choose one)

- updated.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/4186)
<!-- Reviewable:end -->

aymanhab commented:
Not sure if upgrading to swig 4.4 can help, worth trying. If we can upgrade swig instead of changing code that's likely less error prone, eventually we maybe able to retire the swigsimtk hack and use the headers from simbody as is. Wishful thinking...

nickbianco commented:
@aymanhab, as discussed, I now have working builds for Ubuntu 24.04 on CI. Here is a summary of changes that I've made on this branch:

- Upgraded Python version to 3.11 (with NumPy 2.0) on CI.
- Removed "perf" tests on CI. These were a good idea at the time, but rarely used. [I've saved a copy of the existing CI workflow in the opensim-perf repo for future reference](https://github.com/opensim-org/opensim-perf/blob/main/external_workflows/continuous_integration.yml).
- Added Ubuntu 24.04 and Windows 2025 CI jobs. Note that both the Windows 2022 and 2025 jobs use Visual Studio 2022.
- I've updated the Simbody commit and `Mat.h` and `Vec.h` under `Bindings/SWIGSimTK` based on @adamkewley's [recent Simbody changes](https://github.com/simbody/simbody/pull/840). This change is necessary to avoid a warning that newer compilers (e.g., GCC 13) about potential out-of-bounds errors in `SimTK::Mat` constructors. (FYI @adamkewley I had to `#ifndef SWIG` the new `SimTK::Mat` constructors because they weren't playing nice with SWIG and manually add a constructor for `SimTK::Mat33`, which is the only one we expose to the bindings.)
- Converted references to copies in `testOutputReporter.cpp` to avoid dangling reference errors from GCC.
- Changed how data members in `WrapResult` and `PathWrapPoint` are initialized to avoid "possibly uninitialized" errors from GCC.

You've also made several changes to the CI for [supporting PyPI builds](https://github.com/opensim-org/opensim-core/pull/4189). I'm not sure what changes should be merged first. Perhaps we get the PyPI builds working and then we PR in the above changes piecemeal. Let me know what you think makes the most sense.

nickbianco commented:
@aymanhab I've incorporated a few last changes and cleaned up the git history, so this is now ready for review.

nickbianco commented:
The vast majority of our users in the biomechanics community use Windows and Mac, and we haven't actually officially distribute Linux releases of [the main (GUI) application](https://simtk.org/frs/?group_id=91). The only official Linux "release" we have are the conda packages; however, [the Linux packages for the 4.5.2 release do seem to be the most popular](https://anaconda.org/channels/opensim-org/packages/opensim/files).

So maybe there is an argument to start "officially" releasing OpenSim, starting with Ubuntu 24? I'm not totally sure, but having to wait for `std::format` until May 2027 is a bit unfortunate.

> The alternative--which we should definitely investigate--is something like compiling the project in CentOS/AlmaLinux with devtoolset

So this would effectively replace our Ubuntu builds but would allow users to install on more Linux distros? That would definitely be nice.

nickbianco commented:
I would guess the same about the demographics of our Windows/Mac versus Linux users. 

> Using an older OS with backported C++ compilers (usually, CentOS) can be a useful trick in these situations. It can yield binaries that only depend on old libstdc++ symbols. If you're building a library that doesn't have additional system dependencies, you should be able to repackage that binary for Ubuntu, etc.

Ah okay, interesting. That is something I'm not familiar with, but I agree it seems worth considering. Perhaps we could discuss some options after upcoming the OpenSim 4.6 release.
